Principles/6 - Declarative over imperative.md

```diff
@@ -0,0 +1,27 @@
+# Declarative over imperative
+
+Imperative code is often a natural choice for developers who have spent a lot of time in C-style languages. The primitives of many popular programming languages are very easy to get started with writing highly imperative code. However, building imperative APIs can lead to much more error prone systems, more difficult to read code, and it being harder to discern the intention from the implementation for a given piece of logic.
+
+We prefer to write code that exposes a declarative API than an imperative API. A declarative API will let you define what results you would like to see, rather than the way you want them to happen. This type of API leads to simpler code with a higher degree of readability, maintainability, and composability. Where imperative code must exist we often counsel using more declarative and immutable styles within it.
+
+Below are a few examples of decisions that demonstrate this principle, some at the highest level of technology choice, and some at the smallest level of individual function implementations:
+
+* We prefer foundational technologies that are declarative, such as JSX and GraphQL, rather than imperative DOM manipulation and REST calls. On the build tooling side we choose the declarative Webpack over more imperative tools like Gulp.
+
+* We prefer to render a component that encapsulates a series of imperative DOM actions, such as `<EventListener />` or `<ScrollLock />`, than to expose classes or functions that one would call directly.
+
+* We prefer to use or build declarative APIs, such as `<FormState />` or `<JobPoller />` on top of complex state heavy operations such as managing form state, handling graphql requests and caching, and tracking long running jobs.
+
+* We prefer to expose server modules as declarative middleware functions rather than as stateful imperative classes.
```

Comment:
Feels like a weaker example given that we do end up using things that work this way (router), and for individual middlewares there is no choice but to do this. I would rather bring in how we prefer exposing HTTP response values through components in our application (like Redirect), rather than relying on having any server code at all, which by its nature is imperative

Reply:
Let's make a JSX based server ;)